http/tus.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	handler, err := tusd.NewUnroutedHandler(tusd.Config{
+		BasePath:              basePath,
+		StoreComposer:         composer,
+		NotifyCompleteUploads: true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create handler: %s", err)
+	}
+
+	// Goroutine to handle completed uploads
+	go func() {
+		for event := range handler.CompleteUploads {
+			if err := th.handleTusFileUploaded(handler, d, event); err != nil {
+				log.Printf("ERROR: couldn't handle completed upload: %s\n", err)
+			}
+		}
+	}()
+
+	return handler, nil
+}
+
+func getMetadataField(metadata tusd.MetaData, field string) (string, error) {
+	if value, ok := metadata[field]; ok {
+		return value, nil
+	} else {
+		return "", fmt.Errorf("metadata field %s not found in upload request", field)
+	}
+}
+
+func (th tusHandler) handleTusFileUploaded(handler *tusd.UnroutedHandler, d *data, event tusd.HookEvent) error {
+	// Clean up only if an upload has been finalized
+	if !event.Upload.IsFinal {
+		return nil
+	}
+
+	filename, err := getMetadataField(event.Upload.MetaData, "filename")
+	if err != nil {
+		return err
+	}
+	destination, err := getMetadataField(event.Upload.MetaData, "destination")
+	if err != nil {
+		return err
+	}
+	overwriteStr, err := getMetadataField(event.Upload.MetaData, "overwrite")
+	if err != nil {
+		return err
+	}
+	uploadDir := filepath.Join(d.user.FullPath("/"), th.uploadDirName)
+	uploadedFile := filepath.Join(uploadDir, event.Upload.ID)
+	fullDestination := filepath.Join(d.user.FullPath("/"), destination)
+
+	log.Printf("Upload of %s (%s) is finished. Moving file to destination (%s) "+
+		"and cleaning up temporary files.\n", filename, uploadedFile, fullDestination)
+
+	// Check if destination file already exists. If so, we require overwrite to be set
+	if _, err := os.Stat(fullDestination); !errors.Is(err, os.ErrNotExist) {
+		if overwrite, err := strconv.ParseBool(overwriteStr); err != nil {
+			return err
+		} else if !overwrite {
+			return fmt.Errorf("overwrite is set to false while destination file %s exists", destination)
+		}
+	}
+
+	// Move uploaded file from tmp upload folder to user folder
+	if err := os.Rename(uploadedFile, fullDestination); err != nil {
+		return err
+	}
+
+	// Remove uploaded tmp files for finished upload (.info objects are created and need to be removed, too))
+	for _, partialUpload := range append(event.Upload.PartialUploads, event.Upload.ID) {
+		filesToDelete, err := filepath.Glob(filepath.Join(uploadDir, partialUpload+"*"))
+		if err != nil {
+			return err
+		}
+		for _, f := range filesToDelete {
+			if err := os.Remove(f); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Delete folder basePath if it is empty
+	dir, err := ioutil.ReadDir(uploadDir)
+	if err != nil {
+		return err
+	}
+
+	if len(dir) == 0 {
+		// os.Remove won't remove non-empty folders in case of race condition
+		if err := os.Remove(uploadDir); err != nil {
+			return err
+		}
+	}
```

Comment:
This code can be executed after ` os.MkdirAll(uploadDir, os.ModePerm);` but before the actual upload is started